### PR TITLE
system.autoUpgrade: add boot option

### PR DESCRIPTION
**Pull request by Djabx**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/tasks/auto-upgrade.nix [hunk 1]
```diff
@@ -223,7 +234,7 @@ in {
         ''}
 
         if [ "''${booted}" = "''${built}" ]; then
-          ${nixos-rebuild} switch ${toString cfg.flags}
+          ${nixos-rebuild} ${toString (cfg.target)} ${toString cfg.flags}
```
Comment (bjornfor) [inline nixos/modules/tasks/auto-upgrade.nix:237]: The argument is already a string, so `toString` shouldn't be necessary here.

File: nixos/modules/tasks/auto-upgrade.nix [hunk 2]
```diff
@@ -21,6 +21,17 @@ in {
         '';
       };
 
+      target = mkOption {
```
Comment (bjornfor) [inline nixos/modules/tasks/auto-upgrade.nix:24]: Bikeshed: rename `target` to `operation`?

(`colmena` calls this arg `goal`, but that doesn't feel like a good fit when reading the `system.autoUpgrade.*` options IMHO.)

File: nixos/modules/tasks/auto-upgrade.nix [hunk 3]
```diff
@@ -21,6 +21,17 @@ in {
...
+        type = types.str;
```
Comment (bjornfor) [inline nixos/modules/tasks/auto-upgrade.nix:25]: I think this option should be of type `types.enum ["boot" "switch"]` or something.

Comment (Djabx) [reply nixos/modules/tasks/auto-upgrade.nix:25]: Yep, would be great, but so far I don't know how to make taht.

File: nixos/modules/tasks/auto-upgrade.nix (hunk 2 above)
Comment (Djabx) [reply nixos/modules/tasks/auto-upgrade.nix:24]: Ok for me.

File: nixos/modules/tasks/auto-upgrade.nix (hunk 1 above)
Comment (Djabx) [reply nixos/modules/tasks/auto-upgrade.nix]: ok.

File: nixos/modules/tasks/auto-upgrade.nix (hunk 3 above)
Comment (bjornfor) [reply nixos/modules/tasks/auto-upgrade.nix:25]: Just `type = types.enum ["boot" "switch"];` instead of the `type = types.str;` line.

Comment (Djabx) [reply nixos/modules/tasks/auto-upgrade.nix]: Ok done :)

**Review by bjornfor — APPROVED** [review]

**Review by ck3d — APPROVED** [review]

**Review by superherointj — APPROVED** [review]

**Pull request merged**